ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/TransactionService.scala

```diff
@@ -0,0 +1,707 @@
+          contract <- ledger.create(party, Dummy(party))
+        } yield contract))
+        endAfterFirstSection <- ledger.currentEnd()
+        firstSectionRequest = ledger
+          .getTransactionsRequest(Seq(party))
+          .update(_.end := endAfterFirstSection)
+        firstSection <- ledger.flatTransactions(firstSectionRequest)
+        _ <- Future.sequence(Vector.fill(transactionsToSubmit)(for {
+          contract <- ledger.create(party, Dummy(party))
+        } yield contract))
+        endAfterSecondSection <- ledger.currentEnd()
+        secondSectionRequest = ledger
+          .getTransactionsRequest(Seq(party))
+          .update(_.begin := endAfterFirstSection, _.end := endAfterSecondSection)
+        secondSection <- ledger.flatTransactions(secondSectionRequest)
+        fullSequence <- ledger.flatTransactions(party)
+      } yield {
+        val concatenation = Vector.concat(firstSection, secondSection)
+        assert(
+          fullSequence == concatenation,
+          s"The result of processing items in two chunk should yield the same result as getting the overall stream of transactions in the end but there are differences. " +
+            s"Full sequence: ${fullSequence.map(_.commandId).mkString(", ")}, " +
+            s"first section: ${firstSection.map(_.commandId).mkString(", ")}, " +
+            s"second section: ${secondSection.map(_.commandId).mkString(", ")}"
+        )
+      }
+  }
+
+  private[this] val identicalAndParallel = LedgerTest(
+    "TXParallel",
+    "The same data should be served for two identical, parallel requests") { ledger =>
+    val transactionsToSubmit = 5
+    val parallelRequests = 10
+    for {
+      party <- ledger.allocateParty()
+      _ <- Future.sequence(Vector.fill(transactionsToSubmit)(for {
+        contract <- ledger.create(party, Dummy(party))
+      } yield contract))
+      results <- Future.sequence(Vector.fill(parallelRequests)(for {
+        transactions <- ledger.flatTransactions(party)
+      } yield transactions))
+    } yield {
+      assert(
+        results.toSet.size == 1,
+        s"All requests are supposed to return the same results but there " +
+          s"where differences: ${results.map(_.map(_.commandId)).mkString(", ")}"
+      )
+    }
+  }
+
+  private[this] val notDivulgeToUnrelatedParties =
+    LedgerTest("TXNotDivulge", "Data should not be exposed to parties unrelated to a transaction") {
+      ledger =>
+        for {
+          Vector(alice, bob) <- ledger.allocateParties(2)
+          _ <- ledger.create(alice, Dummy(alice))
+          bobsView <- ledger.flatTransactions(bob)
+        } yield {
+          assert(
+            bobsView.isEmpty,
+            s"After Alice create a contract, Bob sees one or more transaction he shouldn't, namely those created by commands ${bobsView.map(_.commandId).mkString(", ")}"
+          )
+        }
+    }
+
+  private[this] val endToEndShouldBeEmpty =
```

Comment:
Super nit-picky, but maybe move this test after the test that uses ledger begin twice?

Reply:
Sure, why not? :slightly_smiling_face:

Reply:
Makes a lot of sense actually. Addressed by https://github.com/digital-asset/daml/pull/2713/commits/cf1d1d348c6cbaa9bd443c51609dd8f736884657